### ISSUE TITLE
Ability to drag around the minefield and zoom in/out

### ISSUE THREAD

**Issue by calhix** [riscygeek/billig-sweeper#5]
For smaller devices, the ability to drag the minefield around when you zoom in is great for larger minefields so you can better ensure you don't misclick, especially when on expert mode/larger fields. Perhaps you could add a flag to enable these options as adding these abilities may not be best on desktop.

**Comment by riscygeek** [comment]
Done. It works on desktops. I'm gonna have to dig out my Pinephone and test it there later...

**Comment by calhix** [comment]
Panning works really well on mobile but may want to disable for desktop as I noticed it's hard to play fast since if your mouse is slightly moving the clicks don't register, or add an option in menu to disable panning.

**Comment by riscygeek** [comment]
I agree, I tried improving the desktop experience by introducing a delay, but that didn't help.
I think I'm just gonna bind it to a keybinding on desktop (probably Space).

**Comment by calhix** [comment]
Cool that works

**Comment by calhix** [comment]
Are you able to set a boundary to how far you can pan? Just noticed you can pan to the point that you can't see the minefield.

**Comment by riscygeek** [comment]
I forgot to add that, thanks for the reminder.

**Comment by calhix** [comment]
May want to consider disable panning on the game over/you win screen.

**Comment by riscygeek** [comment]
Fixed, as of commit d38d2f4

**Comment by calhix** [comment]
Honestly the intermediate preset grid size is perfect, it's about the smallest you can go and be accurate on mobile at least for me. As a stopgap if you could make every preset the same size as the intermediate squares, every config would work great without zoom. Even on desktop intermediate is a good default size for the squares and of course you can zoom how you want with the scroll wheel.

**Comment by riscygeek** [comment]
I'm gonna be looking into how touchscreens work with SDL2 in the next days. **Zooming on mobile will be fixed**.
On desktop, I think I'm gonna map zooming on the middle-mouse button and space.

As for the preset sizes, I don't know. I'm gonna ask around and see what is best, but thanks for the suggestion.
You know what, I'm probably gonna make it even configurable in the config! That's probably best. Maybe even add a "Save as preset" button or alike. But that's not for sure, I'm currently thinking ahead.

**Comment by riscygeek** [comment]
As of 7b1cb4d, presets are now configurable in the config file.

**Comment by riscygeek** [comment]
As of 6c6043b, touch support should be greatly improved.
Currently panning & zooming is implemented, with opening the menu still relying on the keyboard.

**Comment by riscygeek** [comment]
Forgot to mention that long-click is now also implemented, as well as haptic feedback.

**Comment by calhix** [comment]
Great work so far! Excited to try it out

**Comment by riscygeek** [comment]
It works on Mobian Phosh. Should I compile it for you, or can you DIY?

**Comment by calhix** [comment]
I can test later today.

**Comment by riscygeek** [comment]
ok, I'm gonna put as a package in the 1.3-rc1 release.

**Comment by riscygeek** [comment]
Sorry for the delay, flathub had some issues, but you can install the beta version with:
`flatpak install https://dl.flathub.org/build-repo/106336/xyz.stuerz.BilligSweeper.flatpakref`

**Comment by calhix** [comment]
It's really good, zoom is very smooth!

**Comment by riscygeek** [comment]
What do you think about the long-click delay and the haptic feedback intensity/duration. In my opinion the delay and duration could be a little smaller. But what do you think?

**Comment by calhix** [comment]
With the ability now to pan and zoom, I think an always visible taskbar at the top with the time, menu button and mines remaining maybe (that one not a real important function) would be great to avoid requiring the keyboard at all.

**Comment by riscygeek** [comment]
Yeah, I think I'm gonna implement something like this for v1.4.
Probably with the option to turn it off.

**Comment by calhix** [comment]
Sounds good, also a way to pause the game would be nice.

**Comment by calhix** [comment]
Oh and about the delay, I was thinking the same thing, could be slightly shorter. I'm not getting any haptic feedback though.

**Comment by riscygeek** [comment]
It could be a flatpak-permissions related problem. I'm gonna investigate. Build from source it works fine.

**Comment by riscygeek** [comment]
Yeah, it's flatpak-related. If you have Flatseal installed, you can tick on "All Devices" or "device=all". This should make it work.

**Comment by calhix** [comment]
That works! I think the vibration could be shorter as well, and maybe add vibration upon hitting a mine/game over?

**Comment by riscygeek** [comment]
Added a simple death vibration as of 4067e39 and adjusted long-click vibrator settings.
I think I'm gonna release everything with 1.3 later today (It's almost 2 AM where I live).

**Comment by calhix** [comment]
Ah sounds good have a good rest!
